data_structures/priority_queue.cpp

```diff
@@ -0,0 +1,154 @@
+ * @author [Puneet Tripathi](https://github.com/PuneetTri)
+ */
+
+#include<iostream>
+using namespace std;
+
+// Class 
+class PQ{
+    private:
+        int capacity; // max capacity of PQ
+        int *arr = NULL; // Pointer to the array of the PQ
+        int N; // current size of PQ or total elements currently in PQ
+        
+        // Helper function to swap elements
+        void swap(int *a, int *b) {
+            int temp = *a;
+            *a = *b;
+            *b = temp;
+        }
+
+        /* Helper function for the max-heap implementation of PQ
+        *  Function would help promte child node to their correct 
+        *  position
+        */
+        void swim(int pos) {
+            // Check if parent is smaller than child node
+            while(pos > 1 && (arr[pos/2] < arr[pos])) {
+                // In such case swap value of child with parent
+                swap(&arr[pos], &arr[pos/2]);
+                pos = pos/2; // Jump to position of parent node
+            }
+            // Promotion of child node will go on until it becomes smaller than the parent
+        }
+
+        /* Helper function for the max-heap implementation of PQ
+        *  Function would help demote parent node to their correct 
+        *  position
+        */
+        void sink(int pos) {
+            // Check if node's position is that of parent node
+            while(2*pos <= N) {
+                int current = 2*pos; // Jump to the positon of child node
+                // Compare both the children for the greater one
+                if(current < N && arr[current] < arr[current+1]) current++;
+                // If the parent node is greater, sink operation is complete. Break the loop
+                if(arr[pos] >= arr[current]) break; 
+                swap(&arr[pos], &arr[current]); // If not exchange the value of parent with child
+                pos = current; // Exchange parent position to child position in the array
+            }
+        }
+    public:
+        PQ() {
+            /* If capacity is not defined, default size of 11 would be used
+            *  capacity=12 because we cant access 0th element of PQ, and to 
+            *  accomodate 11 elements we need capacity to be 12.
+            *  Parent is at position k, child at position (k*2,k*2+1), if we
+            *  use positon 0 in our queue, its child would be at:
+            *  (0*2, 0*2+1) -> (0,0). This is why we start at position 1
+            */
+            capacity = 12; 
+            arr = new int[capacity];
+            for(int i=0; i<capacity; i++) arr[i] = 0; // Initializing array with value 0
+            N = 0; // Initializing size of PQ 
+        }
+        PQ(int max) {
+            capacity = max+1;
+            arr = new int[capacity];
+            for(int i=0; i<capacity; i++) arr[i] = 0; // Initializing array with value 0
+            N = 0; // Initializing size of PQ 
+        }
+        
```

Comment:
Memory needs to be deleted after is has been acquiered via free - RAII. 

```suggestion
~PQ() { delete[] arr; }       
```